[rush] On errors, have `rush build` print underlying tool's stdout when stderr is empty

hogmoru opened the pull request:
This would fix #1101 :
In case of build failure, if stderr is empty, `rush build` would print stdout instead.
This improves experience when using Webpack & other tools that improperly print their errors to stdout and nothing at all to stderr.

octogonz commented on apps/rush-lib/src/logic/taskRunner/test/TaskRunner.test.ts:
```diff
@@ -108,5 +108,24 @@ describe('TaskRunner', () => {
+        isIncrementalBuildAllowed: false,
+        execute: (writer: ITaskWriter) => {
+          writer.write('Step 1 of important task' + EOL);
+          writer.write('Oh noes' + EOL);
```
Use professional wording

hogmoru replied:
Well, we're testing a tool that prints errors to stdout, which is utterly non-professional, I thought it would be more realistic this way :wink: . Okay, okay, will edit.

octogonz commented on apps/rush-lib/src/logic/taskRunner/TaskRunner.ts:
```diff
@@ -428,13 +428,16 @@ export class TaskRunner {
+          const stderr: string = task.writer.getStdError();
+          const shouldPrintDetails: boolean =
+            task.status === TaskStatus.Failure || task.status === TaskStatus.SuccessWithWarning;
+          let details: string = stderr ? stderr : task.writer.getStdOutput();
```
Will `getStdOutput()` potentially return a much larger amount of text than `stderr`?  If so, should we limit it to a maximum number of lines?

hogmoru replied:
Reasoning was: "since we print all of stderr without limit, might as well do the same with stdout". But this morning I agree it's not great, as stdout is very often considerably larger in practice, so practically speaking it makes sense to make that effort for stdout. Thanks, will see how I can improve.

hogmoru replied:
I'm thinking we might want the first and last N lines here. Do you prefer bespoke code within `TaskRunner.ts`, or should I add a reusable method in `Text.ts`?

octogonz replied:
Generally we tend to keep stuff scoped as privately as possible at first.  We can promote it to being shared (or eventually public API) later when we encounter some other component that actually needs it. The rationale is that if we don't have at least one concrete example of another caller, then we're probably not in a great situation to be designing the API yet.

hogmoru commented:
Thanks for your feedback, so, I've removed frivolities from text samples in tests, and added a method to summarize output if too large. That trimming happens in all cases, whether text comes from tool's stdout or stderr. I arbitrarily decided on leaving first 10 and last 20 lines, is that alright?

Please note a subtle change of behavior, I only noticed while testing that previous code was using `trim()` on all lines, which removes all leading and trailing blanks. I _think_ output indentation matters, so I changed this and now I only remove the trailing blanks. Is that ok?

iclanton commented on apps/rush-lib/src/logic/taskRunner/TaskRunner.ts:
```diff
@@ -443,4 +443,21 @@ export class TaskRunner {
+    const amountRemoved: number = lines.length - headSize - tailSize;
+    const head: string = lines.splice(0, headSize).join(os.EOL);
+    const tail: string = lines.splice(-tailSize).join(os.EOL);
+    return `${head}${os.EOL}[...${amountRemoved} lines elided...]${os.EOL}${tail}`;
```
Huh. I've never heard the word "elided" before. Would "omitted" be better?

hogmoru replied:
Sure, to be honest I did not spontaneously think of it: saw it [there](https://twitter.com/garybernhardt/status/1097307043881308161), and thought for a moment it was more accurate than "omit", but "omit" is fine.
Done.

iclanton commented on apps/rush-lib/src/logic/taskRunner/TaskRunner.ts:
```diff
@@ -443,4 +443,21 @@ export class TaskRunner {
+  /**
+   * Remove trailing blanks, and all middle lines if text is large
+   */
+  private _abridgeTaskReport(text: string): string {
```
Can you add a unit test for this case as well?

hogmoru replied:
I thought the snapshots would capture how leading & trailing blanks are handled, this but it was not the case. So I added a separate test to explicitly verify leading blanks are preserved and trailing blanks are trimmed. Is that what you meant here?

(also added a bunch of snapshots)

iclanton commented:
Added a few comments. It'd be good to have a snapshot test of the output truncation. Other than that, looks good to me.

octogonz commented:
This was published with Rush 5.6.0.

nipunn1313 commented:
My project uses a small python script around typescript.
Unfortunately on a failure, python writes to stderr and typescript writes to stdout (https://github.com/microsoft/TypeScript/issues/12844).

I got stuck for a while figuring out the behavior from #1102. I think it makes more sense personally to either
- output everything
- output only stderr

This behavior of outputting stdout only when stderr is empty was very confusing to debug.
Possibly an option to select between behaviors if both are valuable to different folks. IMO, printing everything seems like the right call.
